### PR TITLE
Get data with blocks, not sliding window

### PR DESCRIPTION
**Pull request by nnistelrooij**
The inputs are also normalized to unit Gaussian distributions.

### CODE REVIEW

File: utils/data.py [hunk 1]
```diff
@@ -47,8 +49,15 @@ def __init__(self, calendar, prices, sales, seq_len=1, horizon=1):
         self.prices = self._sell_prices(calendar, prices)
         self.sales = self._unit_sales(sales)
 
+        # normalize inputs to unit Gaussian distribution
+        self.day = (self.day - 1.4436644) / 6.093091
+        self.snap = (self.snap - 2.029751) / 3.42601
+        self.prices = (self.prices - 2.029751) / 3.42601
+        self.sales = (self.sales - 2.029751) / 3.42601
```
Comment (rusane) [inline utils/data.py:56]: Did you hardcode the mean and std because it takes a long time to compute?

**Review by rusane — APPROVED** [review]

File: utils/data.py (hunk 1 above)
Comment (nnistelrooij) [reply utils/data.py:56]: I wasn't sure about this either.

For the first competition, I also hardcoded it, because the normalization should be the same between the training, validation, and testing datasets. The validation and testing datasets do not have access to the samples of the training dataset, so then the options are either hardcoding or static variables. I opted for hardcoding, but if you prefer static variables, I can change it.

What I wasn't sure about is: Should I use the same mean and std for all the different types of data, or should I use a separate mean and std per type of data. (With type of data, I mean the 9 types of weekdays, weeks, monthdays, months, years, event types, SNAP, prices, and sales). The image data of the first competition was all the same type of data, so then it makes sense to use the same mean and std. But should I use the same mean and std for each type, or for two types like it is now, or have a separate mean and std for each of the 9 types of data?

File: optim.py [hunk 2]
```diff
@@ -67,25 +67,20 @@ def train(model, loader, train_writer, optimizer, criterion, epoch, num_days):
         optimizer    = [Optimizer] optimizer to update the model
         criterion    = [nn.Module] neural network module to compute loss
         epoch        = [int] current iteration over the training data set
-        num_days     = [int] number of days to validate
     """
-    for _ in tqdm(range(len(loader) - num_days), desc=f'Train Epoch {epoch}'):
-        day, items, t = next(loader)
-
+    for day, items, t in tqdm(train_loader, desc=f'Train Epoch {epoch}'):
         # predict
```
Comment (MauiGonzo) [inline optim.py:72]: maybe explain what is returned as prediction

File: optim.py [hunk 3]
```diff
@@ -104,43 +99,39 @@ def validate(model, val_loader, val_writer, criterion, epoch):
...
-        sales = y[..., 0].view(1, -1)
-        targets = t[..., 0].view(1, -1)
+        sales = y[:, :1]
+        targets = t[..., :1]
 
         for day, items, t in tqdm(val_loader, desc=f'Validation Epoch {epoch}'):
-            # DONE - check?
-            # REMARK: sales are not unit figures
-            items[0, -len(sales):, 2] = sales[-seq_len:]
+            # replace actual sales in items with projected sales
```
Comment (MauiGonzo) [inline optim.py:112]: ... in items with last seq_len projected sales

File: optim.py [hunk 4]
```diff
@@ -201,27 +190,30 @@ def forward(self, x, y):
...
 
-    loader = ForecastDataset(calendar, prices, sales.iloc[:, :100], seq_len=seq_len, horizon=horizon)
-    loader = DataLoader(loader)
+    val_dataset = ForecastDataset(calendar, prices, val_sales)
+    val_loader = DataLoader(val_dataset)
 
-    train_writer = TensorBoardWriter('cpu', True, log_dir='runs/train/')
-    val_writer = TensorBoardWriter('cpu', False, log_dir='runs/val/')
+    train_writer = TensorBoardWriter(log_dir='runs/train/')
+    val_writer = TensorBoardWriter(eval_freq=1, log_dir='runs/val/')
```
Comment (MauiGonzo) [inline optim.py:208]: ik heb start voor een main.py gemaakt, deze strings zouden daar in de toekomst in kunnen. de eerste competitie

**Review by MauiGonzo — APPROVED** [review]

File: optim.py (hunk 3 above)
Comment (nnistelrooij) [reply optim.py:112]: Nu dus niet meer. Omdat we de data nu in blokken krijgen i.p.v. een sliding window is `seq_len` altijd gelijk aan 1. Deze regel wordt dan een stuk simpler.

File: utils/data.py (hunk 1 above)
Comment (rusane) [reply utils/data.py:56]: > For the first competition, I also hardcoded it, because the normalization should be the same between the training, validation, and testing datasets. The validation and testing datasets do not have access to the samples of the training dataset, so then the options are either hardcoding or static variables. I opted for hardcoding, but if you prefer static variables, I can change it.

Ah I see, static variables might make the code more readable, so that would have a slight preference for me.

> What I wasn't sure about is: Should I use the same mean and std for all the different types of data, or should I use a separate mean and std per type of data. (With type of data, I mean the 9 types of weekdays, weeks, monthdays, months, years, event types, SNAP, prices, and sales). The image data of the first competition was all the same type of data, so then it makes sense to use the same mean and std. But should I use the same mean and std for each type, or for two types like it is now, or have a separate mean and std for each of the 9 types of data?

A separate mean and std for the 9 types of data makes most sense to me. But `weekdays`, `years`, `months`, and `event_types` are all one-hot vectors so it doesn't really make sense to normalize those I think. Actually I don't think it is necessary to normalize any of the variables/features in `self.day` because these are not continuous. Not sure what kind of data `self.snap` is, but if it is just 0s and 1s then I don't think we need to normalize that either,
